cumulusci/salesforce_api/retrieve_profile_api.py

```diff
@@ -0,0 +1,302 @@
+    "FlowDefinition": FLOWDEFINITION,
+}
+SETUP_ENTITY_QUERY_NAME = "setupEntityAccess"
+
+SOBJECT_TYPE = "CustomObject"
+SOBJECT_QUERY_NAME = "sObject"
+
+CUSTOM_TAB_TYPE = "CustomTab"
+CUSTOM_TAB_QUERY_NAME = "customTab"
+
+
+class RetrieveProfileApi(BaseSalesforceApiTask):
+    def _init_task(self):
+        self.api_version = "58.0"
+        super(RetrieveProfileApi, self)._init_task()
+
+    def _retrieve_existing_profiles(self, profiles: List[str]):
+        query = self._build_query(["Name"], "Profile", {"Name": profiles})
+        result = self._run_query(query)
+
+        existing_profiles = []
+        for data in result["records"]:
+            existing_profiles.append(data["Name"])
+
+        return existing_profiles
+
+    def _run_query(self, query):
+        try:
+            result = self.sf.query(query)
+        except (SalesforceGeneralError, ConnectionError):
+            result = self._run_bulk_query(query)
+        return result
+
+    def _run_bulk_query(self, query):
+        table_name = self._extract_table_name_from_query(query)
+        job = self.bulk.create_query_job(table_name, contentType="JSON")
+
+        batch = self.bulk.query(job, query)
+        self.bulk.wait_for_batch(job, batch)
+        self.bulk.close_job(job)
+        results = self.bulk.get_all_results_for_query_batch(batch)
+        for result in results:
+            result = json.load(salesforce_bulk.util.IteratorBytesIO(result))
+            return {"records": result}
+
+    def _extract_table_name_from_query(self, query):
+        from_index = query.upper().find("FROM")
+        if from_index != -1:
+            query = query[from_index + len("FROM") :].strip()
+            table_name = query.split()[0]
+            return table_name
+        else:
+            raise ValueError("Invalid query format. 'FROM' clause not found.")
+
+    def _build_query(
+        self,
+        columns: List[str],
+        table_name: str,
+        where: Dict[str, Union[str, List[str]]] = None,
+    ) -> str:
+        if not columns:
+            raise ValueError("Columns list cannot be empty")
+        if not table_name:
+            raise ValueError("Table name cannot be empty")
+
+        select_clause = ", ".join(columns)
+
+        where_clause = ""
+        if where:
+            where_list = []
+            for key, value in where.items():
+                if isinstance(value, list):
+                    value = ", ".join([f"'{item}'" for item in value])
+                    condition = f"{key} IN ({value})"
+                else:
+                    condition = f"{key} = '{value}'"
+                where_list.append(condition)
+            where_clause = "WHERE " + " AND ".join(where_list)
+
+        query = f"SELECT {select_clause} FROM {table_name} {where_clause}".strip()
+        return query
+
+    # Retrieve all the permissionable entitites for a set of profiles
+    def _retrieve_permissionable_entities(self, profiles: List[str]):
```

Comment:
Hey @aditya-balachander ,

Can we break the function this way -> 1) one function to build queries, 2) execute queries in parallel, 3) one function to process result.